Feature/common feed

lee1nna opened the pull request:
# Pull Request 요약

> PR에 대한 내용을 작성해주세요.

일반 피드 작업 관련 PR 입니다.

<br>

## 작업 내용

> 작업 내용에 대해서 작성해주세요.

- 일반피드 생성화면 UI 개발
- 이미지 업로드 훅으로 분리
- 일반피드 파일첨부 기능 개발
- 해시태그 컴포넌트 적용
- ConfirmPopup 컴포넌트 개발

<br>

## 참고 사항

> 참고 사항이 있다면 첨부해주세요.
- 일반피드 모달의 좌측영역은 동일하여 ImagePreview 컴포넌트로 분리하였습니다.
- 모달의 단계에 따라 우측영역만 변경되므로 AddImage 컴포넌트와 AddContent 컴포넌트로 단계적으로 처리하였습니다.
- 이미지 업로드 로직은 useImageUpload 훅으로 별도로 분리하여 가독성을 개선하고 AddImage, AddContent 컴포넌트는 UI에 집중할 수 있도록 구성하였습니다.
- 게시 또는 삭제 팝업은 ConfirmPopup 컴포넌트로 공통적으로 구현하였습니다.
![image](https://github.com/user-attachments/assets/13ebb8a6-9680-4776-b883-fe84fd3ca666)


<br>

## 관련 이슈

> closing할 이슈 번호를 작성해주세요.

- Close #20
- Close #22 

<br>

lee1nna committed:
[refact] modify main layout directory structure

lee1nna committed:
[fix] file type input not working on modal issue
- remove aria-hidden="true" in modal background
- remove aria-hidden="true" in modal container
- add role in modal div
- change link to button in ModalMenu
- modify modal scss

lee1nna committed:
[style] add flex mixin scss

lee1nna committed:
[style] add image for common feed modal

lee1nna committed:
[feat] publishing add image step in common feed

lee1nna committed:
[refact] common feed code refactoring
- seperate image upload logic hooks
- add image preview common component
- add flex related mixin
- seperate common feed related type

lee1nna committed:
[feat] add image delete features

lee1nna committed:
[feat] add hash tag in feed content

lee1nna committed:
[feat] add commonfeeddata used to API

lee1nna committed:
[refact] remove eslint warning error
- add setData useEffect dependency array
- change image tag to next/image

lee1nna committed:
[feat] add ConfirmPopup component

lee1nna committed:
[feat] add confirm popup in common feed

lee1nna committed:
[refator] common feed overall refactoring
- common feed popup state intergration
- add common feed popup type
- remove children setState props & apply to update function from parent

seongjin2427 commented on src/app/(MainLayout)/components/CommonFeed/CommonFeed.tsx:
```diff
@@ -0,0 +1,136 @@
+    setCurrentImage,
+    removeImage,
+  } = useImageUpload();
+  const [data, setData] = useState<CommonFeedData>({
```
상태명을 `feedData`로 변경해보는건 어떨까요!?
그냥 `data`는 너무 일반적인 명사라서 어떤 데이터인지 명확하지 않은 것 같아서요!

lee1nna replied:
feature/discussion-feed 브랜치에서 변경 될 예정입니다!

seongjin2427 commented on src/app/(MainLayout)/components/CommonFeed/CommonFeed.tsx:
```diff
@@ -0,0 +1,136 @@
+      ...prev,
+      hashtag: newHashtags,
+    }));
+  };
```
`updateImages`, `updateContents`, `updateHashTags`를 하나의 메서드로 변경해보는건 어떨까요!
하나의 메서드로 합치면 코드 복잡성도 낮출 수 있고, 이후 하위 컴포넌트로 전달할 props의 개수도 적으니
나중에 유지보수하거나 코드 읽기도 더 쉬울 것 같아요!

```ts
  const changeFeedData = (nextFeedData: Partial<CommonFeedData>) => {
    setFeedData((prev) => ({
      ...prev,
      ...nextFeedData,
    }));
  };
```

lee1nna replied:
반영했습니다 😊

seongjin2427 commented on src/app/(MainLayout)/components/AddContent/AddContent.tsx:
```diff
@@ -0,0 +1,131 @@
+  onNext?: () => void;
+  images: FeedImage[];
+  currentImage: FeedImage | null;
+  setCurrentImage: React.Dispatch<SetStateAction<FeedImage | null>>;
```
개인적인 의견입니다!
props로 전달받는 상태 변경 함수를 그대로 사용하여 `React.Dispatch`, `SetStateAction` 타입을 사용하는 것보다,
아래와 같이 상태 변경 함수를 호출하는 메서드를 전달받아
타입을 길게 작성하지 않아도 되는 걸 선호하는 편입니다 하하.

- Before
```ts
import { ChangeEvent, SetStateAction, useEffect, useState } from 'react';

...
const SomeComponent = () => {
  const [currentImage, setCurrentImage] = useSate<FeedImage | null>(null);
  ...
}

type AddContentProps = {
  ... 
  setCurrentImage: React.Dispatch<SetStateAction<FeedImage | null>>;
}
```

- After
```ts
import { ChangeEvent, useEffect, useState } from 'react';
// SetStateAction 타입 제거 가능

const SomeComponent = () => {
  const [currentImage, setCurrentImage] = useSate<FeedImage | null>(null);

  const updateCurrentImage = (nextCurrentImage: FeedImage) => {
    setCurrentImage(nextCurrentImage);
  }
  ...
}

type AddContentProps = {
  ... 
  updateCurrentImage: (nextCurrentImage: FeedImage) => void;
}
```

lee1nna replied:
반영했습니다 😊

seongjin2427 commented on src/app/(MainLayout)/components/Navigation/Navigation.tsx:
```diff
@@ -56,7 +57,9 @@ const Navigation = () => {
       icon: <CommonFeedIcon />,
-      component: <Modal onClose={closeModal}>일반피드로 대체 예정입니다</Modal>,
+      component: (
+        <CommonFeed modalStatus={feedModal} setModalStauts={setFeedModal} />
```
오타가 있습니닷! 😅
`setModalStauts` -> `setModalStatus`

lee1nna replied:
feature/discussion-feed 브랜치에서 변경 될 예정입니다!

seongjin2427 commented on src/app/(MainLayout)/components/CommonFeed/CommonFeed.tsx:
```diff
@@ -0,0 +1,136 @@
+
+  const createCommonFeed = () => {
+    // TODO: 일반피드 작성 서버 API 연동
+    console.log('데이터', data);
```
최종적으로 서버로 생성할 피드 정보를 보낼 때,
base64로 인코딩 된 문자열이 아닌 `File` 객체 그대로 전송해야 합니다.

lee1nna replied:
반영했습니다 😊

seongjin2427 reviewed (approved):
고생 하셨습니다!!

lee1nna committed:
[feat] add discussion feed component
- change ts file name (commonFeed -> feed)
- add discussion subject components
- add AddContents feed type props
- modify commonfeed typo

lee1nna committed:
[feat] add Textarea component
- replace Textarea component
- change document name (@type->type)

lee1nna committed:
[refact] refactoring update discussion feed data function
- modify update discussion feed data using Partial
- add modal component max height

lee1nna committed:
[refact] refactoring update feed data function
- modify refactoring feed data using Partial
- add max-height in modal css

lee1nna committed:
[refactor] apply common feed feedback
- remove setStateAction props
- modify common feed update function using Partial
- change image type to match server request value
